nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/main/java/org/apache/nifi/processors/GeoEnrichIPRecord.java

```diff
@@ -0,0 +1,323 @@
+    )));
+
+    public static final List<PropertyDescriptor> GEO_PROPERTIES = Collections.unmodifiableList(Arrays.asList(
+            GEO_CITY, GEO_ACCURACY, GEO_LATITUDE, GEO_LONGITUDE, GEO_COUNTRY, GEO_COUNTRY_ISO, GEO_POSTAL_CODE
+    ));
+
+    private static final List<PropertyDescriptor> DESCRIPTORS = Collections.unmodifiableList(Arrays.asList(
+            GEO_DATABASE_FILE, READER, WRITER, SPLIT_FOUND_NOT_FOUND, IP_RECORD_PATH, GEO_CITY, GEO_ACCURACY, GEO_LATITUDE,
+            GEO_LONGITUDE, GEO_COUNTRY, GEO_COUNTRY_ISO, GEO_POSTAL_CODE
+    ));
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return DESCRIPTORS;
+    }
+
+    protected volatile RecordReaderFactory readerFactory;
+    protected volatile RecordSetWriterFactory writerFactory;
+    protected boolean splitOutput;
+
+    @Override
+    @OnScheduled
+    public void onScheduled(ProcessContext context) throws IOException {
+        super.onScheduled(context);
+
+        readerFactory = context.getProperty(READER).asControllerService(RecordReaderFactory.class);
+        writerFactory = context.getProperty(WRITER).asControllerService(RecordSetWriterFactory.class);
+        splitOutput = context.getProperty(SPLIT_FOUND_NOT_FOUND).asBoolean();
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if (input == null) {
+            return;
+        }
+
+        FlowFile output = session.create(input);
+        FlowFile notFound = splitOutput ? session.create(input) : null;
+        final DatabaseReader dbReader = databaseReaderRef.get();
+        try (InputStream is = session.read(input);
+             OutputStream os = session.write(output);
+             OutputStream osNotFound = splitOutput ? session.write(notFound) : null) {
+            RecordPathCache cache = new RecordPathCache(GEO_PROPERTIES.size() + 1);
+            Map<PropertyDescriptor, RecordPath> paths = new HashMap<>();
+            for (PropertyDescriptor descriptor : GEO_PROPERTIES) {
+                if (!context.getProperty(descriptor).isSet()) {
+                    continue;
+                }
+                String rawPath = context.getProperty(descriptor).evaluateAttributeExpressions(input).getValue();
+                RecordPath compiled = cache.getCompiled(rawPath);
+                paths.put(descriptor, compiled);
+            }
+
+            String rawIpPath = context.getProperty(IP_RECORD_PATH).evaluateAttributeExpressions(input).getValue();
+            RecordPath ipPath = cache.getCompiled(rawIpPath);
+
+            RecordReader reader = readerFactory.createRecordReader(input, is, getLogger());
+            RecordSchema schema = writerFactory.getSchema(input.getAttributes(), null);
```

Comment:
IMO it's better to pass `reader.getSchema()` as the second argument here rather than `null`, or else the RecordSetWriter can't be configured to Inherit Record Schema.

Also I wonder if we could use the new (as of the schema inference capability) pattern of "process first record. update schema, process the rest" here. Otherwise you have to have the target fields explicitly in your schema, which will prevent folks from being able to infer the schema. If we could add the target RecordPaths to the schema automatically, I think that'd make this a lot more flexible and powerful. Thoughts?

Reply:
What's a good example of that pattern?

Reply:
Going to pass on this based on conversations on the mailing list. I'm not sure we should be starting here with the discussion on dynamic schema updates. There are issues with that sort of thing that could really get some big data teams in trouble if we just enable arbitrary stiching in of new fields.

Reply:
I'm fine with not doing the dynamic schema update, but you should still pass `reader.getSchema()` instead of `null` as the second argument here, otherwise you have to explicitly set the writer's schema rather than being able to inherit it from the reader (which may itself have inferred the schema, so you wouldn't necessarily know what to set as the writer's schema).